### PR TITLE
Redesign/admin page

### PR DESCRIPTION
**Pull request by Keczale**
bbbbbbbb

### CODE REVIEW

File: backend/src/components/User/interface.ts [hunk 1]
```diff
@@ -38,4 +38,11 @@ export interface IUserService {
 	 * @memberof IUserService
 	 */
 	removeUser(id: string): Promise<IUserModel>;
+
+	/**
+	 * @param {string[]} id
+	 * @returns {Promise < string[] >}
```
Comment (maximwow70) [inline backend/src/components/User/interface.ts:44]: Suppose that formatting should be like { Promise<string[]> }

File: backend/src/components/User/service.ts [hunk 2]
```diff
@@ -121,6 +121,38 @@ const UserService: IUserService = {
 			throw new Error(error.message);
 		}
 	},
+
+	/**
+	 * @param {string[]} id
+	 * @returns {Promise < IUserModel >}
```
Comment (maximwow70) [inline backend/src/components/User/service.ts:127]: The same.

File: backend/src/routes/UserRouter.ts [hunk 3]
```diff
@@ -19,6 +19,9 @@ router.post('/update-user', UserComponent.updateUser);
 
 router.delete('/:id', AccessGuard.isAdmin, UserComponent.removeUser);
 
+router.post('/remove-users', AccessGuard.isAdmin, UserComponent.removeSelectedUsers);
+
+
```
Comment (maximwow70) [inline backend/src/routes/UserRouter.ts:24]: Do we really need 2 lines as separator?

File: frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.html [hunk 4]
```diff
@@ -1,28 +1,36 @@
 <div class="users-role-management-panel">
-	<div class="users-role-management-panel__header">
-		<h3>Users</h3>
+	<app-progress-bar
+		*ngIf="usersRequestStatus === pendingStatus"
+		[mode]="'indeterminate'"
+		[theme]=darkTheme>
```
Comment (maximwow70) [inline frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.html:5]: Hm, here should be quotes.

File: frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts [hunk 5]
```diff
@@ -8,57 +8,150 @@ import { AuthFacadeService } from 'src/app/libs/auth/store/auth/auth.facade';
...
 	public users: User[];
 
+	public isSelectedUsersRemoving: boolean;
+
+	public pendingStatus: string = EntityStatus.Pending;
+
+	public headerButtonText: string = 'delete';
+
+	public lightTheme: ComponentTheme = ComponentTheme.Light;
+	public darkTheme: ComponentTheme = ComponentTheme.Dark;
```
Comment (maximwow70) [inline frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts:36]: It's possible to provide enum for view:
`public readonly componentTheme: typeof ComponentTheme = ComponentTheme;`

What do you think about that? It's not necessary changes, but if you agree with me, we can apply them.

File: frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts [hunk 6]
```diff
@@ -8,57 +8,150 @@ import { AuthFacadeService } from 'src/app/libs/auth/store/auth/auth.facade';
...
+			columnName: 'Event Manager',
+			dataName: 'isEventManager'
+		}
+	];
+
 	constructor(
-		private userSearchService: UserSearchService,
 		private usersFacadeService: UsersFacadeService,
-		private authFacadeService: AuthFacadeService
+		private authFacadeService: AuthFacadeService,
```
Comment (maximwow70) [inline frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts:70]: I'm not sure that comma placed at the last line of arguments is a good practice.
What do you think about that?

File: frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts [hunk 7]
```diff
@@ -8,57 +8,150 @@ import { AuthFacadeService } from 'src/app/libs/auth/store/auth/auth.facade';
...
 
-	public get updateInProgress(): boolean {
-		return this.userUpdate.status === EntityStatus.Pending;
-	}
+	public ngOnInit(): void {
 
-	public get foundUsers(): User[] {
-		return this.userSearchService.userSearch(this.users, this.searchText);
-	}
+		this.usersFacadeService.users$.pipe(takeUntil(this.destroy$)).forEach((users: EntityWrapper<User[]>) => {
```
Comment (maximwow70) [inline frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts:75]: I prefer to brake actions per lines.
Like:
_service
.pipe
.forEach_
It's a great way to improve readability.

File: frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts [hunk 8]
```diff
@@ -8,57 +8,150 @@ import { AuthFacadeService } from 'src/app/libs/auth/store/auth/auth.facade';
...
 
-	public toggleEventManagerClaim(isEventManager: boolean, user: User): void {
-		const toggledUser: User = isEventManager ?
-						{...user, claims: [...user.claims, UserClaim.EventManager]} :
-						{...user, claims: user.claims.
-							filter((claim: UserClaim) => claim !== UserClaim.EventManager)};
-		this.usersFacadeService.toggleEventManagerClaim(toggledUser);
+	// we need original selected objects for comparison in table component to render this selections
+	public getOriginalSelectedUserObjects(user: User): User {
+		for (const selectedUser of this.selectedUsers) {
```
Comment (maximwow70) [inline frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts:114]: Is it possible to use forEach in this case?

File: frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts [hunk 9]
```diff
@@ -8,57 +8,150 @@ import { AuthFacadeService } from 'src/app/libs/auth/store/auth/auth.facade';
...
+
+	public get updateInProgress(): boolean {
+		return this.userUpdate.status === EntityStatus.Pending;
+	}
+
+	public toggleEventManagerClaim(event: RowToggleOutput<User>): void {
+		const user: User = event.row;
+		const toggledUser: User = event.state ?
+						{...user, claims: [...user.claims.
+							filter((claim: UserClaim) => claim !== UserClaim.EventManager), UserClaim.EventManager]} :
```
Comment (maximwow70) [inline frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts:135]: Looks like we able to encapsulate filter logic in one place and avoid duplications.

File: frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts [hunk 10]
```diff
@@ -8,57 +8,150 @@ import { AuthFacadeService } from 'src/app/libs/auth/store/auth/auth.facade';
...
+							filter((claim: UserClaim) => claim !== UserClaim.EventManager)};
+		this.usersFacadeService.toggleEventManagerClaim(toggledUser);
+	}
+
+	public onCheck(selected: User[]): void {
+		this.selectedUsers = [...selected];
+	}
+
+	public onHeaderButtonClick(selectedUsers: User[]): void {
+		const selectedUsersId: string[] = [];
```
Comment (maximwow70) [inline frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts:146]: We can use _map_ method here.
Please, avoid _for_.

File: frontend/src/app/libs/admin-role-management/store/users/users.effects.ts [hunk 11]
```diff
@@ -15,25 +14,49 @@ export class UsersEffects {
...
+		)
 	);
 
 	@Effect()
 	public SetEventManagerRole: Observable<Action> = this.actions$.pipe(
 		ofType(UsersActionType.ToggleEventManagerRole),
 		switchMap((action: ToggleEventManagerRole)  => {
-				return this.usersRepository.updateUser(action.payload).pipe(
-					map(() => new ToggleEventManagerRoleSuccess())
+				return this.usersRepository.updateUser(this.authConverter.convertToDto(action.payload)).pipe(
```
Comment (maximwow70) [inline frontend/src/app/libs/admin-role-management/store/users/users.effects.ts:28]: It's a good practice to covert entity in repository methods.

File: frontend/src/app/libs/admin-role-management/store/users/users.effects.ts [hunk 12]
```diff
@@ -15,25 +14,49 @@ export class UsersEffects {
...
+				);
+		})
+	);
+
+	@Effect()
+	public RemoveUser: Observable<Action> = this.actions$.pipe(
+		ofType(UsersActionType.RemoveUser),
+		switchMap((action: RemoveUser)  => {
+				return this.usersRepository.removeUser(this.authConverter.convertToDto(action.payload)).pipe(
+					map((user: UserDto) => new RemoveUserSuccess(this.authConverter.convertFromDto(user))),
```
Comment (maximwow70) [inline frontend/src/app/libs/admin-role-management/store/users/users.effects.ts:40]: And this methods should return converted entities.

File: frontend/src/app/libs/admin-role-management/store/users/users.reducer.ts [hunk 13]
```diff
@@ -47,28 +47,101 @@ export function usersReducer(state: UsersState = initialState, action: UsersActi
...
+				...state,
+				users: {
+					...state.users,
 					value:
-					Boolean(state.users.value) ?
 						[
-							...state.users.value.filter((user: User) => user.id !== action.payload.id),
-							action.payload
-						] :
+							// ...state.users.value.filter((user: User) => user.id !== action.payload.id),
```
Comment (maximwow70) [inline frontend/src/app/libs/admin-role-management/store/users/users.reducer.ts:61]: Comments like that should be removed from any PR.

File: frontend/src/app/libs/common-components-demo/components/progress-bar-demo/progress-bar-demo.component.html [hunk 14]
```diff
@@ -11,6 +12,7 @@
 		<app-demo-state [title]="'Progress 35%'" [isDark]="true">
 			<app-progress-bar
 				[progress]="35"
+				[mode] = "'determinate'"
```
Comment (maximwow70) [inline frontend/src/app/libs/common-components-demo/components/progress-bar-demo/progress-bar-demo.component.html:15]: The best point here will be to move this value to component.

Comment (maximwow70) [reply frontend/src/app/libs/common-components-demo/components/progress-bar-demo/progress-bar-demo.component.html:15]: And suppose that we can encapsulate it in ENUM.

File: frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts (hunk 5 above)
Comment (Keczale) [reply frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts:36]: its a good idea and make our code shorter, I've made such changes to that component ans will apply this further. But previous pattern was as standard before and we have a lot of this implementations in our code.

File: frontend/proxy.conf.json [hunk 15]
```diff
@@ -1,6 +1,6 @@
 {
   "/api": {
-    "target": "https://gandalf-backend.herokuapp.com",
+    "target": "https://gandalf-frontend.herokuapp.com",
```
Comment (clain-atomi) [inline frontend/proxy.conf.json:3]: are you sure you need this change?

File: frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts [hunk 16]
```diff
@@ -8,57 +8,158 @@ import { AuthFacadeService } from 'src/app/libs/auth/store/auth/auth.facade';
...
+
+						userForTable.isEventManager =
+							userForTable.claims.includes(UserClaim.EventManager) ?
+							true :
+							false;
+
+						temporaryUsers.push(userForTable);
+					}
+
+					this.users = temporaryUsers;
```
Comment (clain-atomi) [inline frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts:101]: move to separate method

File: frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts [hunk 17]
```diff
@@ -8,57 +8,158 @@ import { AuthFacadeService } from 'src/app/libs/auth/store/auth/auth.facade';
...
 		});
 	}
 
-	public toggleEventManagerClaim(isEventManager: boolean, user: User): void {
-		const toggledUser: User = isEventManager ?
-						{...user, claims: [...user.claims, UserClaim.EventManager]} :
-						{...user, claims: user.claims.
-							filter((claim: UserClaim) => claim !== UserClaim.EventManager)};
-		this.usersFacadeService.toggleEventManagerClaim(toggledUser);
+	// we need original selected objects for comparison in table component to render this selections
```
Comment (clain-atomi) [inline frontend/src/app/libs/admin-role-management/components/users-role-management-panel/users-role-management-panel.component.ts:116]: remove comment

File: frontend/src/app/libs/admin-role-management/services/users-repository.service.ts [hunk 18]
```diff
@@ -2,23 +2,51 @@ import { Injectable } from '@angular/core';
...
+		delete updatedUser.checked;
+
+		return this.http.post<UserDto>(url, updatedUser);
+	}
+
+	public removeUser(user: User): Observable<UserDto> {
+
+		const userDto: Partial<UserDto> = this.authConverter.convertToDto(user);
+
+		const url: string = this.API_URL + `/users/${userDto._id}`;
```
Comment (clain-atomi) [inline frontend/src/app/libs/admin-role-management/services/users-repository.service.ts:41]: using `user.id` would be better here

**Review by clain-atomi — CHANGES_REQUESTED** [review]

File: frontend/proxy.conf.json (hunk 15 above)
Comment (Keczale) [reply frontend/proxy.conf.json:3]: image
I have taken this url from our chat. i had to restore it after change oh local server

File: frontend/debug.log [hunk 19]
```diff
@@ -0,0 +1,10 @@
+[1104/143403.687:ERROR:directory_reader_win.cc(43)] FindFirstFile: The system cannot find the path specified. (0x3)
```
Comment (clain-atomi) [inline frontend/debug.log:1]: remove this file

**Review by clain-atomi — CHANGES_REQUESTED** [review]

**Pull request merged**